File: 007Restore_Binary_Tree_from_Preorder_Traversal/007Restore_Binary_Tree_from_Preorder_Traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def recoverFromPreorder(self, S: str) -> TreeNode:
        path, pos = list(), 0
        while pos < len(S):
            level = 0
            while S[pos] == '-':
                pos+=1
                level+=1
            value = 0
            while pos < len(S) and S[pos].isdigit() :
                value = value * 10 + ord(S[pos]) - ord('0')
                pos+=1
            node = TreeNode(value)
            if level == len(path):
                if path:
                    path[-1].left = node
            else:
                path = path[:level]
                path[-1].right = node
            path.append(node)
        return path[0]             

    def FromPreorder(self, input: TreeNode) -> str:
        add=""
        if input == None:
            return "N,"
        add += (str(input.val))
        add += ','
        add += self.FromPreorder(input.left)
        add += self.FromPreorder(input.right)
        return add

input_1 = "1-2--3--4-5--6--7";
input_2 = "1-2--3---4-5--6---7";
input_3 = "1-401--349---90--88";
fun = Solution()
root_1 = fun.recoverFromPreorder(input_1)
root_2 = fun.recoverFromPreorder(input_2)
root_3 = fun.recoverFromPreorder(input_3)
print(fun.FromPreorder(root_1))
print(fun.FromPreorder(root_2))
print(fun.FromPreorder(root_3))

File: 007Restore_Binary_Tree_from_Preorder_Traversal/007Restore_Binary_Tree_from_Preorder_Traversal.cpp
#include<string>
#include<iostream>
#include<stack>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };
 
class Solution {
public:
    TreeNode* recoverFromPreorder(string S) {
        stack<TreeNode*> path;
        int pos = 0;
        while (pos < S.size()) {
            int level = 0; // 深度
            while (S[pos] == '-') {
                ++level;
                ++pos;
            }

            int value = 0; // 深度后的值
            while (pos < S.size() && isdigit(S[pos])) {
                value = value * 10 + (S[pos] - '0');
                ++pos;
            }
            TreeNode* node = new TreeNode(value); // 写成节点

            if (level == path.size()) {
                if (!path.empty())
                    path.top()->left = node;
            }
            else {
                while (level != path.size()) {
                    path.pop();
                }
                path.top()->right = node;
            }
            path.push(node);
        }
        while (path.size() > 1) {
            path.pop();
        }
        return path.top();
    }

    string FromPreorder(TreeNode* root) {
        string add{};
        if (root == NULL)
            return string("N,");
        add += to_string(root->val);
        add += ",";
        add += FromPreorder(root->left);
        add += FromPreorder(root->right);
        return add;
    }
};

int main() {
    string input_1 = "1-2--3--4-5--6--7";
    string input_2 = "1-2--3---4-5--6---7";
    string input_3 = "1-401--349---90--88";
    Solution fun;
    TreeNode* root_1 = fun.recoverFromPreorder(input_1);
    TreeNode* root_2 = fun.recoverFromPreorder(input_2);
    TreeNode* root_3 = fun.recoverFromPreorder(input_3);
    cout << "root_1:" << fun.FromPreorder(root_1) << endl;
    cout << "root_2:" << fun.FromPreorder(root_2) << endl;
    cout << "root_3:" << fun.FromPreorder(root_3) << endl;
    return 0;
}